commit 24d89011914119e0c23f86eda2d50ab95162d52e
Author: Ralph Campbell <ralph.campbell@qlogic.com>
Date:   Wed Jun 30 17:15:05 2010 -0700

IB/qib: if qib_init() fails, driver fails to clean up properly
    
If qib_init() fails, the driver fails to free memory, unregister device
files, and unregister with the PCIe framework. The driver will unload
without error but a subsequent driver load will cause the system to panic.
This was found by changing the 7220 code to load the serdes microcode
separately and not installing the microcode file.

INTERNAL: bug 117072

Signed-off-by: Ralph Campbell <ralph.campbell@qlogic.com>

diff --git a/drivers/infiniband/hw/qib/qib_init.c b/drivers/infiniband/hw/qib/qib_init.c
index 8dfa9c6..ef60804 100644
--- a/drivers/infiniband/hw/qib/qib_init.c
+++ b/drivers/infiniband/hw/qib/qib_init.c
@@ -1338,8 +1338,18 @@ static int __devinit qib_init_one(struct pci_dev *pdev,
 
 	if (qib_mini_init || initfail || ret) {
 		qib_stop_timers(dd);
+		flush_scheduled_work();
 		for (pidx = 0; pidx < dd->num_pports; ++pidx)
 			dd->f_quiet_serdes(dd->pport + pidx);
+		if (qib_mini_init)
+			goto bail;
+		if (!j) {
+			(void) qibfs_remove(dd);
+			qib_device_remove(dd);
+		}
+		if (!ret)
+			qib_unregister_ib_device(dd);
+		qib_postinit_cleanup(dd);
 		if (initfail)
 			ret = initfail;
 		goto bail;
